rosetta/services/block_side_effect.go

```diff
@@ -0,0 +1,168 @@
+package services
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/coinbase/rosetta-sdk-go/types"
+	ethcommon "github.com/ethereum/go-ethereum/common"
+
+	"github.com/harmony-one/harmony/core"
+	hmytypes "github.com/harmony-one/harmony/core/types"
+	nodeconfig "github.com/harmony-one/harmony/internal/configs/node"
+	shardingconfig "github.com/harmony-one/harmony/internal/configs/sharding"
+	"github.com/harmony-one/harmony/rosetta/common"
+	"github.com/harmony-one/harmony/shard"
+)
+
+// containsSideEffectTransaction checks if the block contains any side effect operations to report.
+func (s *BlockAPI) containsSideEffectTransaction(
+	ctx context.Context, blk *hmytypes.Block,
+) bool {
+	if blk == nil {
+		return false
+	}
+	return s.hmy.IsCommitteeSelectionBlock(blk.Header()) || !s.hmy.IsStakingEpoch(blk.Epoch()) || blk.NumberU64() == 0
+}
+
+const (
+	// SideEffectTransactionSuffix is use in the transaction identifier for each block that contains
+	// side-effect operations.
+	SideEffectTransactionSuffix = "side_effect"
+	blockHashStrLen             = 64
+)
+
+// getSideEffectTransactionIdentifier fetches 'transaction identifier' for side effect operations
+// for a  given block.
+// Side effects are genesis funds, pre-staking era block rewards, and undelegation payouts.
+// Must include block hash to guarantee uniqueness of tx identifiers.
+func getSideEffectTransactionIdentifier(
+	blockHash ethcommon.Hash,
+) *types.TransactionIdentifier {
+	return &types.TransactionIdentifier{
+		Hash: fmt.Sprintf("%v_%v",
+			blockHash.String(), SideEffectTransactionSuffix,
+		),
+	}
+}
+
+// unpackSideEffectTransactionIdentifier returns the blockHash if the txID is formatted correctly.
+func unpackSideEffectTransactionIdentifier(
+	txID *types.TransactionIdentifier,
+) (ethcommon.Hash, *types.Error) {
+	hash := txID.Hash
+	hash = strings.TrimPrefix(hash, "0x")
+	hash = strings.TrimPrefix(hash, "0X")
+	if len(hash) < blockHashStrLen || string(hash[blockHashStrLen]) != "_" ||
+		hash[blockHashStrLen+1:] != SideEffectTransactionSuffix {
+		return ethcommon.Hash{}, common.NewError(common.CatchAllError, map[string]interface{}{
+			"message": "unknown side effect transaction ID format",
+		})
+	}
+	blkHash := ethcommon.HexToHash(hash[:blockHashStrLen])
+	return blkHash, nil
+}
+
+// getSideEffectTransaction returns the side effect transaction for a block if said block has one.
+// Side effects to reports are: genesis funds, undelegation payouts, permissioned-phase block rewards.
+func (s *BlockAPI) getSideEffectTransaction(
+	ctx context.Context, blk *hmytypes.Block,
+) (*types.Transaction, *types.Error) {
+	if !s.containsSideEffectTransaction(ctx, blk) {
+		return nil, common.NewError(common.TransactionNotFoundError, map[string]interface{}{
+			"message": "no side effect transaction found for given block",
+		})
+	}
+
+	var startingOpIndex *int64
```

Comment:
for normal transactions, is the opIndex same as transaction index in the block?

Reply:
This operation index corresponds to the operation index of a single transaction according to the rosetta spec. 
Here is [operation documentation](https://www.rosetta-api.org/docs/models/Operation.html).